### PR TITLE
Reporting relevant attribute change when various faults are detected

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed?  Examples:
* We need to mark relevant attributes dirty and notify reporting engine when various device faults are detected.

#### Change overview
Add interface to report relevant attribute change when various faults are detected.

Please note, the actual user scenarios of this feature is embedded system. Each embedded platform need to call the corresponding defined delegate API when various faults are detected. No reference implementation on Linux to simulate those device faults.   

#### Testing
How was this tested? (at least one bullet point required)
* No good way to test this feature until any embedded platform support this feature. No reference implementation is implemented on Linux simulation. Regression is covered by CI.

### CODE REVIEW

File: src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp [hunk 1]
```diff
@@ -213,6 +213,69 @@ class GeneralDiagnosticDelegate : public DeviceLayer::ConnectivityManagerDelegat
...
+                    // If General Diagnostics cluster is implemented on this endpoint
+                    MatterReportingAttributeChangeCallback(endpointId, GeneralDiagnostics::Id,
+                                                           GeneralDiagnostics::Attributes::ActiveHardwareFaults::Id);
+                }
+            }
+        }
+    }
+
+    // Get called when the Node detects a radio fault has been raised.
+    void OnHardwareFaultsDetected() override
```
Comment (msandstedt) [inline src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp:239]:
```suggestion
    void OnRadioFaultsDetected() override
```

File: src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp [hunk 2]
```diff
@@ -213,6 +213,69 @@ class GeneralDiagnosticDelegate : public DeviceLayer::ConnectivityManagerDelegat
...
+                    // If General Diagnostics cluster is implemented on this endpoint
+                    MatterReportingAttributeChangeCallback(endpointId, GeneralDiagnostics::Id,
+                                                           GeneralDiagnostics::Attributes::GetActiveRadioFaults::Id);
+                }
+            }
+        }
+    }
+
+    // Get called when the Node detects a network fault has been raised.
+    void OnHardwareFaultsDetected() override
```
Comment (msandstedt) [inline src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp:260]:
```suggestion
    void OnNetworkFaultsDetected() override
```

File: src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp (hunk 1 above)
Comment (yufengwangca) [reply src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp]: Thanks, I found those changes still in my working space, failed to git add them to the commit by mis typing

**Review by msandstedt — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]
LGTM

**Comment by andy31415** [comment]
fast track: change by domain owner, sufficient time passed for all timezone review.

**Pull request merged**